packages/ra-core/src/dataProvider/useUpdate.ts

```diff
@@ -45,74 +55,319 @@ import useMutation, { MutationOptions, Mutation } from './useMutation';
+                context: unknown
+            ) => {
+                if (mutationMode === 'pessimistic') {
+                    // Optimistically update to the new value in getOne
+                    const {
+                        resource: callTimeResource = resource,
+                        id: callTimeId = id,
+                    } = variables;
+                    updateCache({
+                        resource: callTimeResource,
+                        id: callTimeId,
+                        data,
+                    });
+
+                    if (reactMutationOptions.onSuccess) {
+                        return reactMutationOptions.onSuccess(
+                            data,
+                            variables,
+                            context
+                        );
+                    }
+                }
+            },
+            onSettled: (
+                data: RecordType,
+                error: unknown,
+                variables: Partial<UseUpdateParams<RecordType>> = {},
+                context: unknown
+            ) => {
+                const {
+                    resource: callTimeResource = resource,
+                    id: callTimeId = id,
+                } = variables;
+                if (
+                    mutationMode === 'optimistic' ||
+                    mutationMode === 'undoable'
+                ) {
+                    // Always refetch after error or success:
+                    queryClient.invalidateQueries([
+                        callTimeResource,
+                        'getOne',
+                        callTimeId,
+                    ]);
+                    queryClient.invalidateQueries([
+                        callTimeResource,
+                        'getList',
+                    ]);
+                }
+
+                if (reactMutationOptions.onSettled) {
+                    return reactMutationOptions.onSettled(
+                        data,
+                        error,
+                        variables,
+                        context
+                    );
+                }
+            },
+        }
     );
 
-    return [update, state];
+    const mutateAsync = async (
+        variables: Partial<UseUpdateParams<RecordType>> = {},
+        callTimeOptions: MutateOptions<
+            RecordType,
+            unknown,
+            Partial<UseUpdateParams<RecordType>>,
+            unknown
+        > = {}
+    ) => {
+        if (mutationMode === 'pessimistic') {
+            return mutation.mutateAsync(variables, callTimeOptions);
+        }
+
+        const { onSuccess, onSettled, onError } = callTimeOptions;
+        const {
+            resource: callTimeResource = resource,
+            id: callTimeId = id,
+            data: callTimeData = data,
+        } = variables;
+
+        // optimistic update as documented in https://react-query.tanstack.com/guides/optimistic-updates
```

Comment:
The react-query doc advises to do this in onMutate, but in that case it's impossible to fire the success side effects right away. I had to use a wrapper for `mutateAsync` (and therefore `mutate`) instead.

Besides, the wrapper was necessary for the undoable mode, which shouldn't call the `mutate` callback until the action is confirmed.